---
title: "Figure1_RVIS"
output: html_document
date: "2023-09-20"
---

```{r}
library(tidyverse)
library(data.table)
library(ggplot2)
library(R.utils)
```


```{r, echo = FALSE}
setwd("~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS")
dee_monoallelic <- fread("Genes/dee_monoallelic.tsv", header = F)
asd_monoallelic <- fread("Genes/asd_monoallelic.tsv", header = F)
dd_monoallelic <- fread("Genes/dd_monoallelic.tsv", header = F)

setwd("~/Library/CloudStorage/Box-Box/RVIS/data/gene_lists")
clingen_HI <- fread("clingen_HI.txt", header = F)
mgi_essential <- fread("mgi_essential.tsv", header = F)
```

```{r}
gnomad.metrics <- fread("~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS/Data/gnomad.v2.1.1.lof_metrics.by_gene.txt")

mutability <- gnomad.metrics %>% 
  mutate(mutability = mu_mis + mu_syn + mu_lof + mu_lof*1.25) %>%
  rename("Gene" = gene) %>% 
  select(Gene, mutability)
```

# GnomAD
```{r, message = FALSE}
source("~/Documents/Dhinsda Lab/Function/gnomadRVIS.R")
```

```{r, echo = FALSE}
setwd("~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS/Data/")
gnomad.maf <- fread("gnomad_exomes_2.1.1.tsv.gz")
MISSENSE
```

```{r, echo = FALSE}
score_cols = c("rvis_afr", "rvis_amr", "rvis_eas",
               "rvis_asj", "rvis_fin", "rvis_nfe", "rvis_sas")
```

```{r, echo = FALSE}
gnomad.filtered <- gnomad.maf |> 
  filter(Transcript %in% gnomad.metrics$transcript) |> 
  filter(Consequence %in% c(SYN, MISSENSE, LOF)) |> 
  filter(FILTER %in% c("None"))
```

```{r}
CalcXY <- function(df, maf.cutoff) {
  df.x <- df |> 
    group_by(Gene) |> 
    summarise(x = n()) 
  
  df.y <- df |> 
    filter(Consequence %in% c(MISSENSE, LOF)) |> 
    group_by(Gene) |>
    summarise(
      afr_y = sum(AF_afr > maf.cutoff, na.rm = T),
      amr_y = sum(AF_amr > maf.cutoff, na.rm = T),
      asj_y = sum(AF_asj > maf.cutoff, na.rm = T),
      eas_y = sum(AF_eas > maf.cutoff, na.rm = T), 
      fin_y = sum(AF_fin > maf.cutoff, na.rm = T),
      nfe_y = sum(AF_nfe > maf.cutoff, na.rm = T), 
      sas_y = sum(AF_sas > maf.cutoff, na.rm = T),
      popmax_y = sum(AF_popmax > maf.cutoff, na.rm = T)
      )
  
  tallies <- df.x |> 
    left_join(df.y) %>% 
    left_join(mutability, by = "Gene") %>% 
    filter(!is.na(mutability))
  
  return(tallies)
}
```

```{r}
CalcRVIS <- function(df, y.colname) {
  df <- df |> 
    replace(is.na(df), 0)
  
  fm <- as.formula(paste(y.colname, "~", "mutability"))
  model <- lm(fm, data = df)
  rvis.scores <- rstudent(model)
  
  return(rvis.scores)
}
```

```{r}
PrintAUC <- function(df, score_cols, gene.list, maf.threshold = 0.0005) {

  df.tallies <- df |> 
  CalcXY(maf.cutoff = maf.threshold)

  # Calculate RVIS for each ancestry
  df <- df.tallies |> 
    mutate( 
         rvis_afr = CalcRVIS(df.tallies, "afr_y"),
         rvis_eas = CalcRVIS(df.tallies, "eas_y"),
         rvis_asj = CalcRVIS(df.tallies, "asj_y"),
         rvis_nfe = CalcRVIS(df.tallies, "nfe_y"),
         rvis_sas = CalcRVIS(df.tallies, "sas_y"),
         rvis_amr = CalcRVIS(df.tallies, "amr_y"),
         rvis_fin = CalcRVIS(df.tallies, "fin_y")
         )
  
  df_long <- df %>% 
    select(c(Gene, all_of(score_cols))) |> 
    pivot_longer(cols = score_cols, 
                 names_to = "score_type", 
                 values_to = "score_value")
  
  df_long <- df_long |> 
    mutate(gene_list = ifelse(Gene %in% gene.list$V1, 1, 0))
  
  # Compute the ROC curve and AUC for each score
  roc_list <- lapply(unique(df_long$score_type), function(col) {
    df_sub <- df_long %>% filter(score_type == col)
    
    roc_obj <- roc(df_sub$gene_list ~ df_sub$score_value)
    auc_val <- auc(roc_obj)
    
    # Create data frame for AUC
    auc_df <- data.frame(score = col, AUC = auc_val)
    
    # Create data frame for ROC curve
    roc_df <- data.frame(score = col, FPR = 1 - roc_obj$sensitivities, TPR = roc_obj$specificities)
    
    list(roc_obj = roc_obj, auc_df = auc_df, roc_df = roc_df)
  })
  
  # Combine all ROC curves and AUCs
  auc_df <- do.call(rbind, lapply(roc_list, `[[`, "auc_df"))
  roc_df <- do.call(rbind, lapply(roc_list, `[[`, "roc_df"))
  
  auc_df <- auc_df %>% 
    arrange(desc(AUC))
  
  # Perform DeLong test for each pair of ROC curves and store results in a new table
  delong_results <- data.frame()
  for (i in 1:(length(roc_list) - 1)) {
    for (j in (i + 1):length(roc_list)) {
      test_result <- roc.test(roc_list[[i]]$roc_obj, roc_list[[j]]$roc_obj)
      delong_results <- rbind(delong_results, data.frame(
        Score1 = roc_list[[i]]$auc_df$score,
        Score2 = roc_list[[j]]$auc_df$score,
        P_Value = test_result$p.value
      ))
    }
  }
  
  return(list(auc_df = auc_df, roc_df = roc_df, delong_results = delong_results))
}
```

```{r}
# Okabe-Ito palette for 8 categories
colorblind_palette <- c(
  "#E69F00", # orange
  "#56B4E9", # sky blue
  "#009E73", # bluish green
  "#0072B2", # blue
  "#D55E00", # vermillion
  "#CC79A7", # reddish purple
  "#999999"
)
```

# Figure 1a: gnomAD RVIS Curves
```{r}
# 0.0005 threshold
df.tallies <- gnomad.filtered |>
  CalcXY(maf.cutoff = 0.0005) %>% 
  select(-x)

xy <- df.tallies |> 
  pivot_longer(cols = -c(Gene, mutability), names_to = "ancestry", values_to = "y") %>% 
  filter(ancestry %in% c("afr_y", "sas_y", "amr_y","eas_y", "asj_y", "nfe_y", "fin_y")) %>% 
  mutate(ancestry = toupper(gsub("_y", "", ancestry))) %>% 
  rename(Ancestry = "ancestry")

xy$Ancestry <- factor(xy$Ancestry, levels = c("AFR", "SAS", "AMR", "EAS", "ASJ", "NFE", "FIN"))
  
figure_1a <- xy |> 
  ggplot(aes(x=mutability, y=y, col=Ancestry)) +
  geom_point(size = 0.5, alpha = 0.5) + 
  geom_smooth(method = "lm", se = FALSE) +
  coord_cartesian(xlim = c(0, 0.0015), ylim = c(0, 450)) + 
  theme_minimal() + 
  xlab("Mutability") + 
  ylab("Common (MAF>0.05%) functional variants")+
  scale_color_manual(values = colorblind_palette)

figure_1a
```

```{r}
ggsave("figure1a.pdf", plot = figure_1a, width = 8, height = 6)
```

```{r}
dee_results <- PrintAUC(gnomad.filtered, score_cols, dee_monoallelic)
dee_AUC <- dee_results$auc_df %>% 
  rename(`DEE Monoallelic\n (n=94)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
dee_delong <- dee_results$delong_results %>% 
  mutate(Gene_List = "DEE") %>% 
  select(Gene_List, Score1, Score2, P_Value)


dd_results <- PrintAUC(gnomad.filtered, score_cols, dd_monoallelic)
dd_AUC <- dd_results$auc_df %>% 
  rename(`DD Monoallelic\n (n=435)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
dd_delong <- dd_results$delong_results %>% 
  mutate(Gene_List = "DD") %>% 
  select(Gene_List, Score1, Score2, P_Value)


asd_results <- PrintAUC(gnomad.filtered, score_cols, asd_monoallelic)
asd_AUC <- asd_results$auc_df %>% 
  rename(`ASD Monoallelic\n (n=190)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
asd_delong <- asd_results$delong_results %>% 
  mutate(Gene_List = "ASD") %>% 
  select(Gene_List, Score1, Score2, P_Value)


mgi_results <- PrintAUC(gnomad.filtered, score_cols, mgi_essential)
mgi_AUC <- mgi_results$auc_df %>% 
  rename(`Mouse Essential\n (n=2454)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
mgi_delong <- mgi_results$delong_results %>% 
  mutate(Gene_List = "MGI") %>% 
  select(Gene_List, Score1, Score2, P_Value)


HI_results <- PrintAUC(gnomad.filtered, score_cols, clingen_HI)
HI_AUC <- HI_results$auc_df %>% 
  rename(`Haploinsufficient\n (n=360)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
HI_delong <- HI_results$delong_results %>% 
  mutate(Gene_List = "HI") %>% 
  select(Gene_List, Score1, Score2, P_Value)
```

```{r}
gnomAD_RVIS_AUC <- rbind(dee_AUC, dd_AUC, asd_AUC, mgi_AUC, HI_AUC)
gnomAD_RVIS_AUC <- gnomAD_RVIS_AUC %>% 
  rename("Ancestry" = score) %>% 
  mutate(Ancestry = toupper(gsub("rvis_", "", Ancestry)))

gnomAD_RVIS_AUC$Ancestry <- factor(gnomAD_RVIS_AUC$Ancestry, levels = c("AFR", "SAS", "AMR", "EAS", "ASJ", "NFE", "FIN"))
gnomAD_RVIS_AUC$`Gene List` <- factor(gnomAD_RVIS_AUC$`Gene List`, 
                                      levels = c("DEE Monoallelic\n (n=94)", "DD Monoallelic\n (n=435)", "ASD Monoallelic\n (n=190)", "Haploinsufficient\n (n=360)", "Mouse Essential\n (n=2454)"))
```

# Figure 1b: gnomAD RVIS Performances by Ancestry
```{r}
figure_1b <- ggplot(gnomAD_RVIS_AUC, aes(x = `Gene List`, y = AUC, color = Ancestry)) +
  geom_point(position = position_jitter(width = 0.2), size = 3, alpha = 0.5) +
  labs(y = "AUC Scores") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank())+
  scale_color_manual(values = colorblind_palette)+
  ylim(0.6, 0.95)

figure_1b
```

```{r}
ggsave("figure1b.pdf", plot = figure_1b, width = 8, height = 6)
```

```{r}
# Delong Test
delongtest <- rbind(dee_delong, dd_delong, asd_delong, mgi_delong, HI_delong)
delongtest

write.csv(delongtest, "~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS_Final/delongtest_RVIS_gnomAD.csv")
```


# Logistic Regression gnomAD
```{r}
PrintLogRegResults <- function(df, score_cols, gene.list, maf.threshold = 0.0005) {
  
  df.tallies <- df |> 
    CalcXY(maf.cutoff = maf.threshold)

  # Calculate RVIS for each ancestry
  df <- df.tallies |> 
    mutate( 
         rvis_afr = CalcRVIS(df.tallies, "afr_y"),
         rvis_eas = CalcRVIS(df.tallies, "eas_y"),
         rvis_asj = CalcRVIS(df.tallies, "asj_y"),
         rvis_nfe = CalcRVIS(df.tallies, "nfe_y"),
         rvis_sas = CalcRVIS(df.tallies, "sas_y"),
         rvis_amr = CalcRVIS(df.tallies, "amr_y"),
         rvis_fin = CalcRVIS(df.tallies, "fin_y")
         )
  
  df_long <- df %>% 
    select(c(Gene, all_of(score_cols))) |> 
    pivot_longer(cols = score_cols, 
                 names_to = "score_type", 
                 values_to = "score_value")
  
  df_long <- df_long |> 
    mutate(gene_list = ifelse(Gene %in% gene.list$V1, 1, 0))
  
  # Logistic Regression Analysis
  log_reg_results <- lapply(unique(df_long$score_type), function(col) {
    df_sub <- df_long %>% filter(score_type == col)
    
    model <- glm(gene_list ~ score_value, data = df_sub, family = binomial)
    summary_df <- summary(model)$coefficients
    
    # Create a dataframe to store results
    results_df <- data.frame(score = col, Estimate = summary_df["score_value", "Estimate"],
                             Std.Error = summary_df["score_value", "Std. Error"],
                             z.value = summary_df["score_value", "z value"],
                             P.val = summary_df["score_value", "Pr(>|z|)"])
    
    return(results_df)
  })
  
  log_reg_df <- do.call(rbind, log_reg_results)
  
  # Print Logistic Regression Results
  return(log_reg_df)
}
```

```{r}
dee_AUC <- PrintLogRegResults(gnomad.filtered, score_cols, dee_monoallelic, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`DEE Monoallelic (n=94)` = P.val) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

dd_AUC <- PrintLogRegResults(gnomad.filtered, score_cols, dd_monoallelic, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`DD Monoallelic (n=435)` = P.val) %>%
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

asd_AUC <- PrintLogRegResults(gnomad.filtered, score_cols, asd_monoallelic, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`ASD Monoallelic (n=190)` = P.val) %>%  
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

mgi_AUC <- PrintLogRegResults(gnomad.filtered, score_cols, mgi_essential, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`Mouse Essential (n=2454)` = P.val) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

HI_AUC <- PrintLogRegResults(gnomad.filtered, score_cols, clingen_HI, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`Haploinsufficient (n=360)` = P.val) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")
```

```{r}
gnomAD_RVIS_log <- rbind(dee_AUC, dd_AUC, asd_AUC, mgi_AUC, HI_AUC)
gnomAD_RVIS_log <- gnomAD_RVIS_log %>% 
  rename("Ancestry" = score)
gnomAD_RVIS_log
```

```{r}
write.csv(gnomAD_RVIS_log, file = "~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS/gnomAD_log_regression.csv")
```




# UKBiobank
```{r}
data <- read_csv("~/Library/CloudStorage/Box-Box/RVIS/data/UKB/maximally_diverse_ancestry_specific_mafs/new_compiled_mafs.csv")
```

```{r, message = FALSE}
source("~/Documents/Dhinsda Lab/Function/gnomadRVIS.R")
```

```{r}
data <- data[,-1]
```

```{r, echo = FALSE}
score_cols = c("rvis_afr", "rvis_asj", "rvis_eas", "rvis_nfe", "rvis_sas")
```

```{r, echo = FALSE}
data.filtered <- data |> 
  filter(Function %in% c(SYN, MISSENSE, LOF)) %>% 
  mutate(Gene = gsub("'", "", `Gene Name`)) %>% 
  select(-`Gene Name`)
```

```{r}
CalcXY <- function(df, maf.cutoff) {
  df.x <- df |> 
    group_by(Gene) |> 
    summarise(x = n()) 
  
  df.y <- df |> 
    filter(Function %in% c(MISSENSE, LOF)) |> 
    group_by(Gene) |>
    summarise(
      afr_y = sum(afr_maf > maf.cutoff, na.rm = T),
      asj_y = sum(asj_maf > maf.cutoff, na.rm = T), 
      eas_y = sum(eas_maf > maf.cutoff, na.rm = T), 
      nfe_y = sum(nfe_20k_maf > maf.cutoff, na.rm = T), 
      sas_y = sum(sas_maf > maf.cutoff, na.rm = T)
      )
  
  tallies <- df.x |> 
    left_join(df.y) %>% 
    left_join(mutability, by = "Gene") %>% 
    filter(!is.na(mutability))
  
  return(tallies)
}
```

```{r}
CalcRVIS <- function(df, y.colname) {
  
  df <- df |> 
    replace(is.na(df), 0)
  
  fm <- as.formula(paste(y.colname, "~", "mutability"))
  model <- lm(fm, data = df)
  rvis.scores <- rstudent(model)
  
  return(rvis.scores)
}
```

```{r}
PrintAUC <- function(df, score_cols, gene.list, maf.threshold = 0.0005) {

  df.tallies <- df |> 
  CalcXY(maf.cutoff = maf.threshold)

  # Calculate RVIS for each ancestry
  df <- df.tallies |> 
    mutate( 
         rvis_afr = CalcRVIS(df.tallies, "afr_y"),
         rvis_eas = CalcRVIS(df.tallies, "eas_y"),
         rvis_asj = CalcRVIS(df.tallies, "asj_y"),
         rvis_nfe = CalcRVIS(df.tallies, "nfe_y"),
         rvis_sas = CalcRVIS(df.tallies, "sas_y"))
  
  df_long <- df %>% 
    select(c(Gene, all_of(score_cols))) |> 
    pivot_longer(cols = score_cols, 
                 names_to = "score_type", 
                 values_to = "score_value")
  
  df_long <- df_long |> 
    mutate(gene_list = ifelse(Gene %in% gene.list$V1, 1, 0))
  
  # Compute the ROC curve and AUC for each score
  roc_list <- lapply(unique(df_long$score_type), function(col) {
    df_sub <- df_long %>% filter(score_type == col)
    
    roc_obj <- roc(df_sub$gene_list ~ df_sub$score_value)
    auc_val <- auc(roc_obj)
    
    # Create data frame for AUC
    auc_df <- data.frame(score = col, AUC = auc_val)
    
    # Create data frame for ROC curve
    roc_df <- data.frame(score = col, FPR = 1 - roc_obj$sensitivities, TPR = roc_obj$specificities)
    
    list(roc_obj = roc_obj, auc_df = auc_df, roc_df = roc_df)
  })
  
  # Combine all ROC curves and AUCs
  auc_df <- do.call(rbind, lapply(roc_list, `[[`, "auc_df"))
  roc_df <- do.call(rbind, lapply(roc_list, `[[`, "roc_df"))
  
  auc_df <- auc_df %>% 
    arrange(desc(AUC))
  
  # Perform DeLong test for each pair of ROC curves and store results in a new table
  delong_results <- data.frame()
  for (i in 1:(length(roc_list) - 1)) {
    for (j in (i + 1):length(roc_list)) {
      test_result <- roc.test(roc_list[[i]]$roc_obj, roc_list[[j]]$roc_obj)
      delong_results <- rbind(delong_results, data.frame(
        Score1 = roc_list[[i]]$auc_df$score,
        Score2 = roc_list[[j]]$auc_df$score,
        P_Value = test_result$p.value
      ))
    }
  }
  
  return(list(auc_df = auc_df, roc_df = roc_df, delong_results = delong_results))
}

```

```{r}
# Okabe-Ito palette for 5 categories
ukb_colorblind_palette <- c(
  "#E69F00", # orange
  "#56B4E9", # sky blue
  "#0072B2", # blue
  "#D55E00", # vermillion
  "#CC79A7" # reddish purple
)
```

# Figure 1c: UKBiobank RVIS Curves
```{r}
# 0.0005 threshold
df.tallies <- data.filtered |>
  CalcXY(maf.cutoff = 0.0005) %>% 
  select(-x)

xy <- df.tallies |> 
  pivot_longer(cols = -c(Gene, mutability), names_to = "ancestry", values_to = "y")%>% 
  mutate(ancestry = toupper(gsub("_y", "", ancestry))) %>% 
  rename(Ancestry = "ancestry")

xy$Ancestry <- factor(xy$Ancestry, levels = c("AFR", "SAS", "EAS", "ASJ", "NFE"))

figure_1c <- xy |> 
  ggplot(aes(x=mutability, y=y, col=Ancestry)) +
  geom_point(size = 0.5, alpha = 0.5) + 
  geom_smooth(method = "lm", se = FALSE) + 
  coord_cartesian(xlim = c(0, 0.0015), ylim = c(0, 450)) + 
  theme_minimal() + 
  xlab("Mutability") + 
  ylab("Common (MAF>0.05%) functional variants")+
  scale_color_manual(values = ukb_colorblind_palette)

figure_1c
```

```{r}
ggsave("figure1c.pdf", plot = figure_1c, width = 8, height = 6)
```

```{r}
dee_results <- PrintAUC(data.filtered, score_cols, dee_monoallelic)
dee_AUC <- dee_results$auc_df %>% 
  rename(`DEE Monoallelic\n (n=94)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
dee_delong <- dee_results$delong_results %>% 
  mutate(Gene_List = "DEE") %>% 
  select(Gene_List, Score1, Score2, P_Value)


dd_results <- PrintAUC(data.filtered, score_cols, dd_monoallelic)
dd_AUC <- dd_results$auc_df %>% 
  rename(`DD Monoallelic\n (n=435)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
dd_delong <- dd_results$delong_results %>% 
  mutate(Gene_List = "DD") %>% 
  select(Gene_List, Score1, Score2, P_Value)


asd_results <- PrintAUC(data.filtered, score_cols, asd_monoallelic)
asd_AUC <- asd_results$auc_df %>% 
  rename(`ASD Monoallelic\n (n=190)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
asd_delong <- asd_results$delong_results %>% 
  mutate(Gene_List = "ASD") %>% 
  select(Gene_List, Score1, Score2, P_Value)


mgi_results <- PrintAUC(data.filtered, score_cols, mgi_essential)
mgi_AUC <- mgi_results$auc_df %>% 
  rename(`Mouse Essential\n (n=2454)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
mgi_delong <- mgi_results$delong_results %>% 
  mutate(Gene_List = "MGI") %>% 
  select(Gene_List, Score1, Score2, P_Value)


HI_results <- PrintAUC(data.filtered, score_cols, clingen_HI)
HI_AUC <- HI_results$auc_df %>% 
  rename(`Haploinsufficient\n (n=360)` = AUC) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "AUC")
HI_delong <- HI_results$delong_results %>% 
  mutate(Gene_List = "HI") %>% 
  select(Gene_List, Score1, Score2, P_Value)
```

```{r}
UKBiobank_RVIS_AUC <- rbind(dee_AUC, dd_AUC, asd_AUC, mgi_AUC, HI_AUC)
UKBiobank_RVIS_AUC <- UKBiobank_RVIS_AUC %>% 
  rename("Ancestry" = score) %>% 
  mutate(Ancestry = toupper(gsub("rvis_", "", Ancestry)))

UKBiobank_RVIS_AUC$Ancestry <- factor(UKBiobank_RVIS_AUC$Ancestry, levels = c("AFR", "SAS", "EAS", "ASJ", "NFE"))
UKBiobank_RVIS_AUC$`Gene List` <- factor(UKBiobank_RVIS_AUC$`Gene List`, 
                                      levels = c("DEE Monoallelic\n (n=94)", "DD Monoallelic\n (n=435)", "ASD Monoallelic\n (n=190)", "Haploinsufficient\n (n=360)", "Mouse Essential\n (n=2454)"))
```

# Figure 1d: UKBiobank RVIS Performances by Ancestry
```{r}
figure_1d <- ggplot(UKBiobank_RVIS_AUC, aes(x = `Gene List`, y = AUC, color = Ancestry)) +
  geom_point(position = position_jitter(width = 0.2), size = 3, alpha = 0.5) +
  labs(y = "AUC Scores") +
  scale_fill_manual(values = c("Ancestry A" = "blue", "Ancestry B" = "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank())+
  scale_color_manual(values = ukb_colorblind_palette)+
  ylim(0.6, 0.95)


figure_1d
```

```{r}
ggsave("figure1d.pdf", plot = figure_1d, width = 8, height = 6)
```

```{r}
# Delong Test
delongtest <- rbind(dee_delong, dd_delong, asd_delong, mgi_delong, HI_delong)
delongtest

write.csv(delongtest, "~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS_Final/delongtest_RVIS_UKB.csv")
```


```{r}
library(patchwork)

patch <- (((figure_1a + theme(axis.title.x = element_text(margin = margin(t = -80, unit = "pt")))) | figure_1b) / ((figure_1c + theme(axis.title.x = element_text(margin = margin(t = -80, unit = "pt"))))  | figure_1d)) +
  plot_annotation(tag_levels = 'A')

patch
```

```{r}
ggsave("figure1.pdf", plot = patch, width = 15, height = 10)
```


# UKB
```{r}
PrintLogRegResults <- function(df, score_cols, gene.list, maf.threshold = 0.0005) {
  
  df.tallies <- df |> 
    CalcXY(maf.cutoff = maf.threshold)

  # Calculate RVIS for each ancestry
  df <- df.tallies |> 
    mutate( 
         rvis_afr = CalcRVIS(df.tallies, "afr_y"),
         rvis_eas = CalcRVIS(df.tallies, "eas_y"),
         rvis_asj = CalcRVIS(df.tallies, "asj_y"),
         rvis_nfe = CalcRVIS(df.tallies, "nfe_y"),
         rvis_sas = CalcRVIS(df.tallies, "sas_y")
         )
  
  df_long <- df %>% 
    select(c(Gene, all_of(score_cols))) |> 
    pivot_longer(cols = score_cols, 
                 names_to = "score_type", 
                 values_to = "score_value")
  
  df_long <- df_long |> 
    mutate(gene_list = ifelse(Gene %in% gene.list$V1, 1, 0))
  
  # Logistic Regression Analysis
  log_reg_results <- lapply(unique(df_long$score_type), function(col) {
    df_sub <- df_long %>% filter(score_type == col)
    
    model <- glm(gene_list ~ score_value, data = df_sub, family = binomial)
    summary_df <- summary(model)$coefficients
    
    # Create a dataframe to store results
    results_df <- data.frame(score = col, Estimate = summary_df["score_value", "Estimate"],
                             Std.Error = summary_df["score_value", "Std. Error"],
                             z.value = summary_df["score_value", "z value"],
                             P.val = summary_df["score_value", "Pr(>|z|)"])
    
    return(results_df)
  })
  
  log_reg_df <- do.call(rbind, log_reg_results)
  
  # Print Logistic Regression Results
  return(log_reg_df)
}
```

```{r}
dee_AUC <- PrintLogRegResults(data.filtered, score_cols, dee_monoallelic, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`DEE Monoallelic (n=94)` = P.val) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

dd_AUC <- PrintLogRegResults(data.filtered, score_cols, dd_monoallelic, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`DD Monoallelic (n=435)` = P.val) %>%
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

asd_AUC <- PrintLogRegResults(data.filtered, score_cols, asd_monoallelic, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`ASD Monoallelic (n=190)` = P.val) %>%  
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

mgi_AUC <- PrintLogRegResults(data.filtered, score_cols, mgi_essential, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`Mouse Essential (n=2454)` = P.val) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")

HI_AUC <- PrintLogRegResults(data.filtered, score_cols, clingen_HI, maf.threshold = 0.0005) %>% 
  select(score, P.val) %>% 
  rename(`Haploinsufficient (n=360)` = P.val) %>% 
  pivot_longer(cols = -1, names_to = "Gene List", values_to = "P Value")
```

```{r}
UKB_RVIS_log <- rbind(dee_AUC, dd_AUC, asd_AUC, mgi_AUC, HI_AUC)
UKB_RVIS_log <- UKB_RVIS_log %>% 
  rename("Ancestry" = score)
UKB_RVIS_log
```

```{r}
write.csv(UKB_RVIS_log, file = "~/Library/CloudStorage/Box-Box/Dhindsa Lab/RVIS/UKB_log_regression.csv")
```


